### ISSUE TITLE
wine: Call from 00006FFFFF463EF7 to unimplemented function ucrtbase.dll.crealf, aborting

### ISSUE THREAD

**Issue by vkuznetsovgn** [webcomics/pywine#65]
Hey!

I'm trying to use scipy library in pywine. My host system is Ubuntu 24.04.

I'm doing the following steps:

```
$ docker run -it --rm tobix/pywine:3.12
$ wine pip install scipy
$ wine python
>>> from scipy.signal import butter
```

Then I'm just importing scipy and getting a huge error:

```
>>> from scipy.signal import butter
wine: Call from 00006FFFFF463EF7 to unimplemented function ucrtbase.dll.crealf, aborting
wine: Unimplemented function ucrtbase.dll.crealf called at address 00006FFFFF463EF7 (thread 00e4), starting debugger...
WineDbg attached to pid 0000
Unhandled exception: unimplemented function ucrtbase.dll.crealf called in 64-bit code (0x006fffff463ef7).
Register dump:
 rip:00006fffff463ef7 rsp:00007ffffe2ea320 rbp:00006ffffb3fa580 eflags:00000202 (   - --  I   - - - )
 rax:00007ffffe2ea368 rbx:00006fffffa7f144 rcx:00007ffffe2ea340 rdx:0000000000000000
 rsi:00007ffffe2ea410 rdi:00006ffffb4044e0  r8:00006ffffeb69c61  r9:00007ffffe2ea418 r10:00007437bea94d70
 r11:00007ffffe2ea588 r12:0000000000000070 r13:0000000000000000 r14:00007437bea94d70 r15:0000000000000000
```

Do you know how to fix it and if it's possible to fix it?

**Comment by TobiX** [comment]
crealf is only a [stub](https://gitlab.winehq.org/wine/wine/-/blob/master/dlls/ucrtbase/ucrtbase.spec#L2253) in wine, which means it's not implemented. Every code using that function WILL crash. scipy uses that function [a lot](https://github.com/scipy/scipy/blob/6cd6872860417a7dcfb6697edd9d55ac92aa1f6c/scipy/linalg/_matfuncs_expm.c#L1745), so it's only natural it crashes under wine.

I don't consider this container anything that should be able to run every piece of Python software. It's mainly intended as a tool for CI builds (for example with PyInstaller).

You might want to raise this issue with either wine or scipy to "fix" this. In both cases, providing a patch would probably provide the fastest way to getting this fixed. Looking at [what the function is supposed to do](https://en.cppreference.com/w/c/numeric/complex/creal) the implementation might be quite simple...

**Comment by TobiX** [comment]
On a more positive note: Thanks for trying my container in new and interesting ways :laughing: - A shame it didn't work out for you.

Thinking more about it I would have expected a sensible compiler to inline such simple functions... But MSVC apparently is no sensible compiler and doesn't really "support" complex numbers anyways (https://learn.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-170, example: https://godbolt.org/z/1WPq3K9hr)

**Comment by TobiX** [comment]
PS: It seems Wine is slowly adding those functions: https://gitlab.winehq.org/wine/wine/-/merge_requests/7225

**Comment by domdfcoding** [comment]
I am seeing this with numpy as well, when importing (directly or indirectly) `numpy.dtypes`. If anyone else has the same problem it only seems to affect numpy 2.2.2 and later - older versions work fine for me.

**Comment by hdy111** [comment]
> 我在导入（直接或间接）时也在 numpy 中看到了这一点。如果其他人有同样的问题，它似乎只影响 numpy 2.2.2 及更高版本 - 旧版本对我来说很好。`numpy.dtypes`

very good    you help me     I success!!

**Comment by TobiX** [comment]
This should now work on `:latest` and `:3.14` thanks to the recently released [Wine 11.0](https://gitlab.winehq.org/wine/wine/-/releases/wine-11.0)
